scylla/src/transport/session.rs

```diff
@@ -1103,10 +1118,20 @@ impl Session {
         let mut last_error: Option<QueryError> = None;
 
         'nodes_in_plan: for node in query_plan {
+            let span = trace_span!("Executing query", node = node.address.to_string().as_str());
```

Comment:
In this function you set the parent span for all `trace!` messages, but what about messages in other functions, called from this one?

Reply:
Ah, you wrapped futures returned by them with `.instrument()`.

Reply:
Yup. I could still potentially miss local synchronous helper functions that way, but there aren't any that I'm aware of. And unfortunately this instrumentation gets quite ugly because we're using nested named loops here, so I can't just wrap everything in a single instrumentation.

Reply:
I think you can try to move out the `'same_node_retries` loop into an async block or a separate function and make it return `Result<RetryDecision>`. The `match retry_decision` block would be just outside the span.

As for the synchronous functions, we have `retry_session.decide_should_retry` which is synchronous and may run custom user logic, which in turn can print trace messages, too.

Reply:
Hmm, now I think that `Result<RetryDecision>` will be insufficient because we won't be able to return on success... So maybe something like `Result<std::ops::ControlFlow<Result<ResT, QueryError>, RetryDecision>` (ugh, that's a mouthful, but the type inference should make it unnecessary to type it out)?

Reply:
let's move it to a follow-up, I'd rather not mess with the flow in this particular patch (in order to make regressions less likely)